lib/v8.js

```diff
@@ -20,9 +20,68 @@ const {
   Serializer: _Serializer,
   Deserializer: _Deserializer
 } = internalBinding('serdes');
+const {
+  ERR_V8_HEAPSNAPSHOT
+} = require('internal/errors').codes;
 const { copy } = internalBinding('buffer');
 const { objectToString } = require('internal/util');
 const { FastBuffer } = require('internal/buffer');
+const { toPathIfFileURL } = require('internal/url');
+const { validatePath } = require('internal/fs/utils');
+const { toNamespacedPath } = require('path');
+const {
+  createHeapSnapshotStream,
+  triggerHeapSnapshot
+} = internalBinding('heap_utils');
+const { Readable } = require('stream');
+const { owner_symbol } = require('internal/async_hooks').symbols;
+const {
+  kUpdateTimer,
+  onStreamRead,
+} = require('internal/stream_base_commons');
+const kHandle = Symbol('kHandle');
+
+
+function writeHeapSnapshot(filename) {
+  if (filename !== undefined) {
+    filename = toPathIfFileURL(filename);
+    validatePath(filename);
+    filename = toNamespacedPath(filename);
+  }
+  return triggerHeapSnapshot(filename);
+}
+
+class HeapSnapshotStream extends Readable {
+  constructor(handle) {
+    super({ autoDestroy: true });
+    this[kHandle] = handle;
+    handle[owner_symbol] = this;
+    handle.onread = onStreamRead;
+  }
+
+  _read() {
+    if (this[kHandle])
+      this[kHandle].readStart();
+  }
+
+  _destroy() {
+    // Release the references on the handle so that
+    // it can be garbage collected.
+    this[kHandle][owner_symbol] = undefined;
+    this[kHandle] = undefined;
+  }
+
+  [kUpdateTimer]() {
+    // Does nothing
+  }
+}
+
+function getHeapSnapshot() {
+  const handle = createHeapSnapshotStream();
+  if (!handle)
```

Comment:
Is this something that we should throw a public error for? I think this is more like a defensive assertion, so using `internal/assert` is good enough?